hummingbot/connector/budget_checker.py

```diff
@@ -0,0 +1,290 @@
+        """
+        Fills in the collateral token and collateral amount fields for the order candidate.
+        If there is insufficient assets to cover the collateral, the order amount is adjusted.
+
+        See the doc string for `adjust_candidate` to learn more about how the adjusted order
+        amount is derived.
+
+        :param order_candidate: The candidate order to check and adjust.
+        :param all_or_none: Should the order amount be set to zero on insufficient balance.
+        :return: The adjusted order candidate.
+        """
+        adjusted_candidate = self.adjust_candidate(
+            order_candidate, all_or_none
+        )
+        self._lock_available_collateral(adjusted_candidate)
+        return adjusted_candidate
+
+    def adjust_candidate(
+        self, order_candidate: OrderCandidate, all_or_none: bool = True
+    ) -> OrderCandidate:
+        """
+        Fills in the collateral token and collateral amount fields for the order candidate.
+
+        If there is insufficient collateral to cover the proposed order amount and
+        the `all_or_none` parameter is set to `False`, the order amount will be adjusted
+        to the greatest amount that the remaining collateral can provide for. If the parameter
+        is set to `True`, the order amount is set to zero.
+
+        :param order_candidate: The candidate order to check and adjust.
+        :param all_or_none: Should the order amount be set to zero on insufficient balance.
+        :return: The adjusted order candidate.
+        """
+        adjusted_candidate = self.populate_collateral_fields(order_candidate)
+        available_collateral = self._exchange.get_available_balance(adjusted_candidate.collateral_token)
+        if adjusted_candidate.collateral_amount < available_collateral:
+            if all_or_none:
+                adjusted_candidate.amount = Decimal("0")
+                adjusted_candidate.collateral_amount = Decimal("0")
+            else:
+                adjusted_candidate = self._adjust_amounts(adjusted_candidate, available_collateral)
+        return adjusted_candidate
+
+    def populate_collateral_fields(self, order_candidate: OrderCandidate) -> OrderCandidate:
+        """
+        Populates the required collateral amount and the collateral token fields for the given order.
+
+        This implementation assumes a spot-specific configuration for collaterals (i.e. the quote
+        token for buy orders and base for sell). It can be overridden to provide other
+        configurations such as in the case of perpetual connectors.
+
+        :param order_candidate: The candidate order to check and adjust.
+        :return: The adjusted order candidate.
+        """
+        trading_pair = order_candidate.trading_pair
+
+        base, quote = split_hb_trading_pair(trading_pair)
+        collateral_token = quote if TradeType.BUY else base
+
+        required_collateral = order_candidate.amount * order_candidate.price
```

Comment:
Shouldn't this be different for buys and sells? If I want to buy 2 BTC with order price 60000 USDT, that means the collateral for the order is 120000 USDT. But if I want to sell 2 BTC with order price 60000 USDT the collateral for the order is 2 BTC, not 2 BTC * 60000 USDT/BTC